src/lib/HST/Effect/InputModule.hs

```diff
@@ -0,0 +1,143 @@
+{-# LANGUAGE BlockArguments #-}
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TemplateHaskell #-}
+
+-- | This module defines an effect that allows to get modules and module
+--   interfaces by the file path and module interfaces by the module name.
+module HST.Effect.InputModule
+  ( -- * Module Interface
+    ModuleInterface(..)
+  , ConEntry(..)
+  , TypeName
+  , ConName
+  , revertInterfaceEntry
+    -- * Effect
+  , InputModule
+    -- * Actions
+  , getInputModule
+  , getInputModuleInterface
+  , getInputModuleInterfaceByName
+    -- * Interpretations
+  , runInputModule
+  ) where
+
+import           Data.Map.Strict     ( Map )
+import qualified Data.Map.Strict     as Map
+import           Data.Maybe          ( mapMaybe )
+import           Polysemy            ( Member, Sem, interpret, makeSem )
+
+import           HST.Effect.Report   ( Report, reportFatal )
+import qualified HST.Frontend.Syntax as S
+import           HST.Util.Messages   ( Severity(Error), message )
+
+-------------------------------------------------------------------------------
+-- Aliases for Names in Module Interfaces                                    --
+-------------------------------------------------------------------------------
+-- | The name of a data type.
+type TypeName a = S.QName a
+
+-- | The name of a data constructor.
+type ConName a = S.QName a
+
+-------------------------------------------------------------------------------
+-- Module Interface Data Types                                               --
+-------------------------------------------------------------------------------
+-- | A data type for module interfaces that stores the name and the data types
+--   of a module.
+data ModuleInterface a = ModuleInterface
+  { interfaceModName   :: Maybe (S.ModuleName a)
+    -- ^ The name of the module or @Nothing@, if the module does not have a
+    --   name.
+  , interfaceDataCons  :: Map (TypeName a) [ConEntry a]
+    -- ^ A map that maps data type names to their constructors.
+  , interfaceTypeNames :: Map (ConName a) (TypeName a)
+    -- ^ A map that maps constructor names to the names of their data types.
```

Comment:
Before we had

```haskell
envConEntries  :: Map (ConName a) (ConEntry a)
envDataEntries :: Map (TypeName a) (DataEntry a)
```

which correspond to `interfaceTypeNames` and `interfaceDataCons` with the only difference being the missing `DataEntry` and `ConEntry` wrappers. I think that we should add the wrappers back, such that it becomes easier to store more information about data types and constructors in the future.

```suggestion
interfaceConEntries  :: Map (ConName a) (ConEntry a)
interfaceDataEntries :: Map (TypeName a) (DataEntry a)
```

Reply:
Only the `DataEntry` type has been removed, not the `ConEntry` type. The `DataEntry` type has become quite useless after changing the types of these maps. However, if anyone wants to add additional information about data types in the future, it wouldn't be difficult to add such a wrapper back in.
And the main change in these maps, that they are now "crossed" (type names map to constructors and constructor names map to types) was done deliberately to make lookups more efficient for our purposes. If you had a constructor name and wanted to get the constructor entries of the type that this constructor belongs to (which actually is the only case we currently use these lookup functions for), you had to:
1. Get the constructor entry of the constructor name with a lookup in `interfaceConEntries`.
2. Extract the type name of the constructor entry.
3. Lookup this type name in `interfaceDataEntries` to get the data entry.
4. Extract the constructor names of the data entry.
5. Lookup the constructor entry for every single constructor name you want to have the constructor entry for in `interfaceConEntries`.

With the new map types, you have to:
1. Get the type name of the constructor name with a lookup in `interfaceTypeNames`.
2. Lookup this type name in `interfaceDataCons` to get all constructor entries.

The only downside to this implementation is the one I mentioned in the PR description: If you only want the single constructor entry for one constructor name, then the old maps are better.